### DISCUSSION TITLE
Misidentifying code as CLIPS

### DISCUSSION THREAD

**Discussion by JetForMe** [github-linguist/linguist#6252] [Classification] [answered]
## Describe the bug

Does Linguist rely primarily on filename extension? It seems to, despite the docs talking about heuristics and other methods.

Looking at my repo (a LinuxCNC set of configuration and code files), Linguist has misidentified at least two files. [One](https://github.com/JetForMe/router-table/blob/master/fusion360/LZAvidCNCRotary.cps) ends in `.cps`, which it identifies as Component Pascal, but is actually JavaScript, and the [other](https://github.com/JetForMe/router-table/blob/master/linuxcnc/config/router3/ladder.clp) ends in `.clp`, which it identifies as CLIPS, but is actually more of a configuration file format used by LinuxCNC for ladder logic.

### Expected behaviour

I suppose it's a reasonable way to go, so feel free to close this "bug," but I wanted to point it out. The file contents look nothing like the Lisp of CLIPS.

In any case, I'm working on adding a `.gitattributes` file to help correct this for this repo, although that only seems to work for the branches in which it exists, despite that information applying generally across the repo.

**Comment by lildude** [comment, marked as answer]
> Does Linguist rely primarily on filename extension? It seems to, despite the docs talking about heuristics and other methods.

Linguist uses multiple strategies as detailed in https://github.com/github/linguist/blob/master/docs/how-linguist-works.md It works from most specific to least specific in an attempt to reduce the number of possible languages as it progresses through the strategies whilst still being efficient. The extension does however have a large influence as it is one of the first strategies used and if that extension is only associated with one language, that language is used and no further stragies are considered.

> Looking at my repo (a LinuxCNC set of configuration and code files), Linguist has misidentified at least two files. [One](https://github.com/JetForMe/router-table/blob/master/fusion360/LZAvidCNCRotary.cps) ends in `.cps`, which it identifies as Component Pascal, but is actually JavaScript, and the [other](https://github.com/JetForMe/router-table/blob/master/linuxcnc/config/router3/ladder.clp) ends in `.clp`, which it identifies as CLIPS, but is actually more of a configuration file format used by LinuxCNC for ladder logic.

This will be because the `.cps` extension is only associated with one language:

https://github.com/github/linguist/blob/c1c34e5260797b4d598f5ec76f19723bfc5a1894/lib/linguist/languages.yml#L1221-L1231

The same applies to the `.clp` extension:

https://github.com/github/linguist/blob/c1c34e5260797b4d598f5ec76f19723bfc5a1894/lib/linguist/languages.yml#L797-L804

Support can be added for other languages, but usage levels across GitHub need to be met.

> I suppose it's a reasonable way to go, so feel free to close this "bug," but I wanted to point it out. The file contents look nothing like the Lisp of CLIPS.

Yes, this is expected and be design. Looking at file content is an expensive operation so is only done when no other simpler strategies identify a language.

> In any case, I'm working on adding a `.gitattributes` file to help correct this for this repo, although that only seems to work for the branches in which it exists, despite that information applying generally across the repo.

Linguist only analyses the default branch. This is [documented](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-repository-languages).
